darshan-util/pydarshan/darshan/backend/cffi_backend.py

```diff
@@ -657,3 +658,82 @@ def _log_get_heatmap_record(log):
     libdutil.darshan_free(buf[0])
     
     return rec
+
+
+def log_get_derived_metrics(log_path: str, mod_name: str):
+    """
+    Returns the darshan_derived_metrics struct from CFFI/C accumulator code.
+
+    Parameters:
+        log_path: Path to the darshan log file
+        mod_name: The name of the module to retrieve derived metrics for
+
+    Returns:
+        darshan_derived_metrics struct (cdata object)
+    """
+    # TODO: eventually add support for i.e., a regex filter on the records
+    # the user wants to get derived metrics for--like filtering to records
+    # with a single filename involved before accumulating the data?
+    log_handle = log_open(log_path)
+    jobrec = ffi.new("struct darshan_job *")
+    libdutil.darshan_log_get_job(log_handle['handle'], jobrec)
+    modules = log_get_modules(log_handle)
+
+    if mod_name not in modules:
+        raise ValueError(f"{mod_name} is not in the available log file "
+                         f"modules: {modules.keys()}")
+
+    mod_type = _structdefs[mod_name]
+    darshan_accumulator = ffi.new("darshan_accumulator *")
+    r = libdutil.darshan_accumulator_create(modules[mod_name]['idx'],
+                                            jobrec[0].nprocs,
+                                            darshan_accumulator)
+    if r != 0:
+        raise RuntimeError("A nonzero exit code was received from "
+                           "darshan_accumulator_create() at the C level. "
+                           f"This could mean that the {mod_name} module does not "
+                           "support derived metric calculation, or that "
+                           "another kind of error occurred. It may be possible "
+                           "to retrieve additional information from the stderr "
+                           "stream.")
+
+    buf = ffi.new("void **")
+    r = 1
+    while r >= 1:
+        r = libdutil.darshan_log_get_record(log_handle['handle'], modules[mod_name]['idx'], buf)
```

Comment:
This code is explicitly opening the log and iterating through the records to produce the derived metrics.  The hope with the accumulator API, though, was that it could be applied (kind of like a filter) on records during the first/main pass rather than going back and doing another pass over the log.  The calculations done by the accumulator functions are modest compared to the cost of retrieving records.

I don't know what performance impact that has in practice, though, or how awkward it would be from a code organization point of view.

Reply:
The easiest thing for me would be to have the records read in to a single contiguous buffer that I could then pass around. I think we've discussed that before.

As it stands, we add the records 1-by-1 to pandas Dataframes for counters and fcounters, but the C API doesn't support/accept Python objects holding the records. Writing the accumulator logic to work directly on on our record (pandas) dataframes would allow:

1) Not reading the data in again
2) Easily filtering the dataframe (records) using regexes in Python/pandas, etc.
3) Maintaining the analyses at a higher level
4) The ability to use GPUs/concurrency on large record dataframes with API-matched pandas backends like: https://github.com/rapidsai/cudf 

I'm not sure we'll come to an agreement on the path forward though in the short term. My suggestion would be to aim for correctness now and efficiency later. Note that we probably want to review the base code in the parent PR rather than this extension for file tables.

What I don't want is to basically end up having to write a bunch of C-like code at the CFFI level just to use the API more efficiently--if we're using dataframes to store the records I think long-term we should produce single dataframes for counters and fcounters, and then do any work on them directly. Since the logic is already in C, we're avoiding the duplication for now, but with all the work the community puts into accelerated dataframe operations, I'm not so sure we gain much long-term with hand-writing filters vs. the accelerator backends popping up.

The only problem with this is that we need to dump everything in to memory/buffer, which is also something we've discussed before.

Reply:
Ok, got it.  That's a good synopsis of the issue that helps clarify what's going on.  Agree with the strategy to get correctness now and then circle back to it.  Do you want copy that synopsis into an issue for how to optimize the aggregate metrics?

Reply:
FWIW, I'm fine just targeting correctness here, as well.

I manually compared time to generate reports for the E3SM-IO test cases (one with DXT data, one without)  on this branch and compared to `main`, and I don't see any obvious slowdown, though it's clearly costing more to make 2 passes over log data. Guess that means other aspects of report generation are dominating overheads.

I was thinking we could potentially add some shims to CFFI bindings / report interface to automatically add records to an accumulator as they are read in (i.e., before we fully convert them to our Python format and free the memory associated with the C-style format). That would keep us at one pass, and we could maintain the accumulators and derived metrics within the report API so other tools could use them. However, that kind of restricts the generality of accumulators at the PyDarshan level (users couldn't just create an accumulator for a subset of files they are interested in, they would only be implicitly maintained on a per-module basis). If we want to keep the accumulators more general (as intended), we'd still need to solve the general problem of converting from the Python format back to what the C library expects...

Reply:
Looking closer at the code here, I think we're at 3 total passes over the Darshan log, though. One for the traditional job summary pass we've always one done, one for per-module bandwidth calc, and one for the per-module file count table. Could we just cache the derived metrics and aggregate record data returned from emit for each module rather than doing that twice?

Reply:
`log_get_derived_metrics()` is the shared entrypoint if that is what you mean. We could try adding an [`@lru_cache`](https://docs.python.org/3/library/functools.html#functools.lru_cache) decorator, as long as the hashability requirements are met. I think if you grep the Python part of the codebase you'll see a few other uses of that construct.

I think we're on the same page re: Python vs. C workflow awkwardness. If you want to make it nice and convenient to filter the records prior to accumulator/derived metrics processing at the Python level, it looks messy to me at the moment re: regex application to a C data structure instead of `pandas`, etc.